crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -0,0 +1,302 @@
+                perpetual_db,
+                /* num_versions_per_object */ 3,
+                /* num_object_versions_to_retain */ 2,
+                /* total_unique_object_ids */ 1000,
+            )
+            .await;
+            // We had 3 versions for 1000 unique object ids, we wanted to retain 2 version per object id i.e. we should have pruned 1000 keys
+            assert_eq!(1000, total_pruned.unwrap());
+        }
+        {
+            let primary_path = tempfile::tempdir()?.into_path();
+            let perpetual_db = Arc::new(AuthorityPerpetualTables::open(&primary_path, None));
+            // add 3 versions for every object id and 1000 unique object ids
+            let total_pruned = test_pruning(
+                perpetual_db,
+                /* num_versions_per_object */ 2,
+                /* num_object_versions_to_retain */ 2,
+                /* total_unique_object_ids */ 1000,
+            )
+            .await;
+            // We had only 2 version per key, we did not prune anything
+            assert_eq!(0, total_pruned.unwrap());
+        }
+        {
+            let primary_path = tempfile::tempdir()?.into_path();
+            let perpetual_db = Arc::new(AuthorityPerpetualTables::open(&primary_path, None));
+            // add 3 versions for every object id and 1000 unique object ids
+            let total_pruned = test_pruning(
+                perpetual_db,
+                /* num_versions_per_object */ 1,
+                /* num_object_versions_to_retain */ 2,
+                /* total_unique_object_ids */ 1000,
+            )
+            .await;
+            // We had only 1 version per key, we did not prune anything
+            assert_eq!(0, total_pruned.unwrap());
+        }
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_db_size_after_compaction() -> Result<(), anyhow::Error> {
+        let primary_path = tempfile::tempdir()?.into_path();
+        let perpetual_db = Arc::new(AuthorityPerpetualTables::open(&primary_path, None));
+        let total_unique_object_ids = 100_000;
+        let num_versions_per_object = 10;
+        let ids = ObjectID::in_range(ObjectID::ZERO, total_unique_object_ids)?;
+        for id in ids {
+            for i in (0..num_versions_per_object).rev() {
+                perpetual_db.objects.insert(
+                    &ObjectKey(id, SequenceNumber::from(i)),
+                    &Object::immutable_with_id_for_testing(id),
+                )?;
+            }
+        }
+        perpetual_db.objects.rocksdb.flush()?;
+
+        let before_compaction_size = get_size(primary_path.clone()).unwrap();
+
+        let total_pruned = AuthorityStorePruner::prune_objects(2, perpetual_db.clone());
+        info!("Total pruned keys = {:?}", total_pruned);
+        let start = ObjectKey(ObjectID::ZERO, SequenceNumber::MIN);
+        let end = ObjectKey(ObjectID::MAX, SequenceNumber::MAX);
+        perpetual_db.objects.compact_range(&start, &end)?;
```

Comment:
how frequent will compaction be triggered in prod?

Reply:
I think we are planning to default to run this (full pruning) once a day for now on both validators and full nodes.

Reply:
oh I mean the things that `compact_range` does.

Reply:
oh sorry i misread your comment. Auto compaction runs frequently on different files+levels in different levels of the LSM tree. We default to level compaction where a new compaction is triggered once the number of files (sst file sizes are fixed) overflows in a level starting from L0 to all the way down.

Reply:
Also, `compact_range` is manual compaction which we don't do today and just rely on auto compaction to be good enough in doing its job.